File: May 30 day challenge/Week 3/Odd even linked list.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def createLinkedList(self,nodeToForm,nodeToCopy):
        while nodeToCopy is not None:
            nodeToForm.val = nodeToCopy.val
            if nodeToCopy.next is not None and nodeToCopy.next.next is not None:
                nodeToForm.next = ListNode()
            else:
                break
            nodeToForm = nodeToForm.next
            nodeToCopy = nodeToCopy.next.next
        return nodeToForm
    def oddEvenList(self, head: ListNode) -> ListNode:
        if not head:
            return head
        newHead = ListNode()
        dup = head
        dup2 = newHead
        newHead = self.createLinkedList(newHead, head)
        if dup.next is None:
            newHead.next = None
            return dup2
        newHead.next = ListNode()
        newHead = self.createLinkedList(newHead.next, dup.next)
        newHead.next = None
        return dup2
head = ListNode()
dup = head
# for num in range(1,6):
#     head.val = num
#     if num != 5:
#         head.next = ListNode()
#     head = head.next
obj = Solution().oddEvenList(dup)
while obj is not None:
    print(obj.val, end=" ")
    obj = obj.next

File: December 30 day challenge 2021/OddEvenLinkedList.java
import java.io.*;
import java.util.*;

public class OddEvenLinkedList {
    public ListNode oddEvenList(ListNode head) {
        if(head == null || head.next == null)
            return head;
        ListNode odd = head, even = head.next, temp = even;
        while(odd.next != null && odd.next.next != null || even.next != null && even.next.next != null){
            if(odd.next != null && odd.next.next != null){
                odd.next = odd.next.next;
                odd = odd.next;
            }
            if(even.next != null && even.next.next != null){
                even.next = even.next.next;
                even = even.next;
            }
        }
        odd.next = temp;
        even.next = null;
        return head;
    }
}
